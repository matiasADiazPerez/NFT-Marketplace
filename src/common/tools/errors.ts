import { HttpException, HttpStatus } from '@nestjs/common';

export class InvalidId extends Error {
  constructor() {
    super('Invalid Id');
  }
}

export class DeletedEntity extends Error {
  constructor(entity: string) {
    super(`The ${entity} is deleted`);
  }
}

export class EntityNotFound extends Error {
  constructor(entity: string, id: number) {
    super(`Could not find the ${entity} with id ${id}`);
  }
}

export const HandleErr = (
  err: InvalidId | DeletedEntity | HttpException | Error,
) => {
  if (err instanceof InvalidId || err instanceof DeletedEntity) {
    throw new HttpException(err.message, HttpStatus.BAD_REQUEST);
  }

  if (err instanceof EntityNotFound) {
    throw new HttpException(err.message, HttpStatus.NOT_FOUND);
  }

  if (err instanceof HttpException) {
    throw err;
  }
  throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
};
