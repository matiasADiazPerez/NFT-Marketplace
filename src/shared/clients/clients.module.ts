import { Global, Module } from '@nestjs/common';
import { Client } from './clients.service';

@Global()
@Module({
  exports: [Client],
  providers: [Client],
})
export class ClientsModule {}
