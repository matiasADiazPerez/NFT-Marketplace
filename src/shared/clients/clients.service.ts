import { Injectable } from '@nestjs/common';
import { ThirdwebSDK } from '@thirdweb-dev/sdk';
import { Sepolia } from '@thirdweb-dev/chains';
import { ethers } from 'ethers';
import { MetaMaskInpageProvider } from '@metamask/providers';

declare global {
  interface Window {
    ethereum?: MetaMaskInpageProvider;
  }
}

@Injectable()
export class ClientsService extends ThirdwebSDK {
  constructor() {
    super(Sepolia, {
      secretKey:
        'E0BlJpurpl66Sm8LYS7nmBZ3kQI4hL26ViE-Rrnd8Sx1rquMbFZLl8AjdIecKSt5IKK_nlmXukC__jbAkaeEKg',
    });
  }
  public async mint(): Promise<any> {
    try {
      const provider = new ethers.providers.Web3Provider(window.ethereum)
      // It will prompt user for account connections if it isnt connected
      const signer = provider.getSigner();
      console.log('Account:', await signer.getAddress());
      const contract = await this.getContract(
        '0xFCE9b92eC11680898c7FE57C4dDCea83AeabA3ff',
      );
      console.log(contract.abi);
      const ans = await contract.call('mint', [
        '0xAd2A6C33946410EeF6Dc0e21b6CF1089D821C623',
      ]);
      return ans;
    } catch (err) {
      console.log(err);
      throw err;
    }
  }
}
