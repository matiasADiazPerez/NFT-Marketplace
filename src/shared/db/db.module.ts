import { Global, Module } from '@nestjs/common';
import { Db } from './db.service';

@Global()
@Module({
  exports: [Db],
  providers: [Db],
})
export class DbModule {}
