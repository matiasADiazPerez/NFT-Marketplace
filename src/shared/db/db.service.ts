import { Injectable } from '@nestjs/common';
import { Bid } from 'models/bids.entity';
import { SellOffer } from 'models/sellOffer.entity';
import { EntityType as ET } from 'models/entityType.interface';
import { User } from 'models/users.entity';

@Injectable()
export class Db {
  bids: Bid[];
  sellOffers: SellOffer[];
  users: User[];

  constructor() {
    this.bids = [];
    this.sellOffers = [];
    this.users = [];
  }

  private _lastIndex<T>(db: Array<T>) {
    if (db.length === 0) {
      return 0;
    }
    return db.length - 1;
  }

  private _getEntity<T extends ET>(id: number, db: T[]): T {
    if (id > this._lastIndex(db)) {
      throw new Error('Invalid id');
    }
    const entity = db[id];
    if (entity?.DeletedAt) {
      throw new Error('The sellOffer is deleted');
    }
    return entity;
  }

  getBid(id: number): Bid {
    return this._getEntity(id, this.bids);
  }
  getSellOffer(id: number): SellOffer {
    return this._getEntity(id, this.sellOffers);
  }
  getUser(id: number): User {
    return this._getEntity(id, this.users);
  }
  lastBidIndex(): number {
    return this._lastIndex(this.bids);
  }
  lastSellIndex(): number {
    return this._lastIndex(this.sellOffers);
  }
  lastUserIndex(): number {
    return this._lastIndex(this.users);
  }
}
