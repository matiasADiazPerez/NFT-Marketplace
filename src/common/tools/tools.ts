import { Client } from 'src/shared/clients/clients.service';
import { AuctionState, SellOffer } from 'models/sellOffer.entity';
import { User } from 'models/users.entity';
import { utils } from 'web3';

export const finishAuction = async (
  seller: User,
  bidder: User,
  offer: SellOffer,
  client: Client,
) => {
  const bid =
    offer.auctionState === AuctionState.OnSale ? offer.price : offer.currentBid;
  const weiBid = utils.toWei(bid.toString(), 'ether');
  const buyInput = {
    tokenId: offer.tokenId.toString(),
    bid: weiBid,
  };

  const approveToken = client.approveToken(
    bidder.userKey,
    process.env.FINISH_CONTRACT,
    buyInput.bid,
  );
  const approveNft = client.approveNft(
    seller.userKey,
    process.env.FINISH_CONTRACT,
    buyInput.tokenId,
  );
  await Promise.all([approveToken, approveNft]);
  const ans = await client.finishAuction(
    bidder.userKey,
    seller.userKey,
    buyInput,
  );
  if (ans['transactionHash']) {
    offer.finishAuctionTx = String(ans['transactionHash']);
  }
  offer.auctionState = AuctionState.Closed;
};
