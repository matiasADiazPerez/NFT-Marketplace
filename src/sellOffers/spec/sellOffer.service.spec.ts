import { Test, TestingModule } from '@nestjs/testing';
import { HttpException, HttpStatus } from '@nestjs/common';
import { SellOffersService } from '../sellOffer.service';
import { Db } from 'src/shared/db/db.service';
import { Client } from 'src/shared/clients/clients.service';
import { User } from 'models/users.entity';
import { DeletedEntity, InvalidId, NotOwner } from 'src/common/tools/errors';
import { AuctionState, SellOffer } from 'models/sellOffer.entity';
import * as toSpyModule from 'src/common/tools/tools';

enum SellOfferId {
  auction,
  deleted,
  onSale,
  toClose,
  toCancel,
}

enum UsersIds {
  seller,
  bidder,
}

const testUser = new User({
  password: 'testPass',
  entropy: 'test entropy',
});

const testBidder = new User({
  password: 'testPass2',
  entropy: 'test entropy2',
});

const testSellOffer = new SellOffer({ tokenId: 0 }, 0);

const deletedTestSellOffer = new SellOffer({ tokenId: 0 }, 0);
deletedTestSellOffer.DeletedAt = new Date();

const onSaleOffer = new SellOffer({ tokenId: 0, price: 100 }, 0);

const toCloseOffer = new SellOffer({ tokenId: 0 }, 0);
toCloseOffer.highestBidder = UsersIds.bidder;

const toCancel = new SellOffer({ tokenId: 10 }, 0);

const client = new Client();
jest.mock('src/shared/clients/clients.service');
jest.mock('src/common/tools/tools');
describe('SellOffersService', () => {
  let service: SellOffersService;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: SellOffersService,
          useFactory: () => {
            const db = new Db();
            db.users.push(testUser);
            db.users.push(testBidder);
            db.sellOffers.push(testSellOffer);
            db.sellOffers.push(deletedTestSellOffer);
            db.sellOffers.push(onSaleOffer);
            db.sellOffers.push(toCloseOffer);
            const nftService = new SellOffersService(client, db);
            return nftService;
          },
        },
      ],
    }).compile();

    service = module.get<SellOffersService>(SellOffersService);
  });
  describe('Create a sell offer', () => {
    it('should be defined', () => {
      expect(service).toBeDefined();
    });
    it('should create a sell offer', async () => {
      const mockOwnerOf = jest
        .spyOn(Client.prototype, 'ownerOf')
        .mockResolvedValue(testUser.userAddr);
      const res = await service.create(0, { tokenId: 0 });
      expect.assertions(2);
      expect(mockOwnerOf).toBeCalledWith('0');
      expect(res).toBe(4);
    });
    it('should fail if the user does not owns the nft', async () => {
      jest
        .spyOn(Client.prototype, 'ownerOf')
        .mockResolvedValue('wrong address');

      const expectedErr = new HttpException(
        'You are not the owner of the nft',
        HttpStatus.FORBIDDEN,
      );
      try {
        expect.assertions(1);
        await service.create(0, { tokenId: 0 });
      } catch (err) {
        expect(err).toMatchObject(expectedErr);
      }
    });
    it('should gracefully fail if the db fails', async () => {
      jest
        .spyOn(Client.prototype, 'ownerOf')
        .mockResolvedValue(testUser.userAddr);
      const invalid = new InvalidId();
      const expectedErr = new HttpException(
        invalid.message,
        HttpStatus.BAD_REQUEST,
      );
      try {
        expect.assertions(1);
        await service.create(2, { tokenId: 0 });
      } catch (err) {
        expect(err).toMatchObject(expectedErr);
      }
    });
  });
  describe('find all sell offers', () => {
    it('should return all sell offers', () => {
      const res = service.findAll();
      expect(res).toMatchObject([testSellOffer, onSaleOffer, toCloseOffer]);
    });
  });
  describe('find offers owned by user', () => {
    it('should return all sell offers of the user', () => {
      const res = service.findByUserId(0);
      expect(res).toMatchObject([
        testSellOffer,
        deletedTestSellOffer,
        onSaleOffer,
        toCloseOffer,
      ]);
    });
  });
  describe('find one sell offers', () => {
    it('should return the corresponding sell offer', () => {
      const res = service.findOne(SellOfferId.auction);
      expect(res).toMatchObject(testSellOffer);
    });
  });
  describe('update price of offer', () => {
    it('should update the price of corresponding sell offer', () => {
      const res = service.updatePrice(0, SellOfferId.onSale, { price: 110 });
      expect(res.price).toBe(110);
    });
    it('should fail if sell offer is deleted', () => {
      try {
        expect.assertions(1);
        service.updatePrice(0, SellOfferId.deleted, { price: 110 });
      } catch (err) {
        const deletedErr = new DeletedEntity('sellOffer');
        const expectedErr = new HttpException(
          deletedErr.message,
          HttpStatus.BAD_REQUEST,
        );
        expect(err).toMatchObject(expectedErr);
      }
    });

    it('should fail if user is not onwer of sell offer', () => {
      try {
        expect.assertions(1);
        service.updatePrice(1, SellOfferId.onSale, { price: 110 });
      } catch (err) {
        expect(err instanceof NotOwner).toBeTruthy();
      }
    });
    it('should fail if the sell offer is not on sale', () => {
      try {
        expect.assertions(1);
        service.updatePrice(0, SellOfferId.auction, { price: 110 });
      } catch (err) {
        const expectedErr = new HttpException(
          'The offer can not be updated',
          HttpStatus.CONFLICT,
        );
        expect(err).toMatchObject(expectedErr);
      }
    });
  });
  describe('remove sell offer', () => {
    it('should remove a sell offer', () => {
      service.remove(0, SellOfferId.auction);
      const sellOffers = service.findAll();
      expect(sellOffers).toMatchObject([onSaleOffer, toCloseOffer]);
    });
    it('should fail if user is not onwer of sell offer', () => {
      try {
        expect.assertions(1);
        service.remove(1, SellOfferId.onSale);
      } catch (err) {
        expect(err instanceof NotOwner).toBeTruthy();
      }
    });
    it('should gracefully fail if the db fails', async () => {
      const invalid = new InvalidId();
      const expectedErr = new HttpException(
        invalid.message,
        HttpStatus.BAD_REQUEST,
      );
      try {
        expect.assertions(1);
        service.remove(0, 6);
      } catch (err) {
        expect(err).toMatchObject(expectedErr);
      }
    });
  });
  describe('close an auction', () => {
    it('should close an auction', async () => {
      const mockFinisAuction = jest
        .spyOn(toSpyModule, 'finishAuction')
        .mockResolvedValue();
      await service.close(0, SellOfferId.toClose);
      expect(mockFinisAuction).toBeCalledWith(
        testUser,
        testBidder,
        toCloseOffer,
        client,
      );
      expect(toCloseOffer.auctionState).toBe(AuctionState.Closed);
    });
    it('should fail if user is not onwer of sell offer', async () => {
      try {
        expect.assertions(1);
        await service.close(1, SellOfferId.toClose);
      } catch (err) {
        expect(err instanceof NotOwner).toBeTruthy();
      }
    });
    it('should fail if is not auction', async () => {
      try {
        expect.assertions(1);
        await service.close(0, SellOfferId.onSale);
      } catch (err) {
        const expectedErr = new HttpException(
          'The offer can not be closed now',
          HttpStatus.CONFLICT,
        );
        expect(err).toMatchObject(expectedErr);
      }
    });
    it('should gracefully fail if the db fails', async () => {
      try {
        expect.assertions(1);
        await service.close(0, 5);
      } catch (err) {
        const invalid = new InvalidId();
        const expectedErr = new HttpException(
          invalid.message,
          HttpStatus.BAD_REQUEST,
        );
        expect(err).toMatchObject(expectedErr);
      }
    });
  });
  describe('cancel a sell offer', () => {
    it('Should cancel a sellOffer', () => {
      service.cancel(0, SellOfferId.toClose);
      expect(toCloseOffer.auctionState).toBe(AuctionState.Closed);
    });
    it('should fail if user is not onwer of sell offer', () => {
      try {
        expect.assertions(1);
        service.cancel(1, SellOfferId.toCancel);
      } catch (err) {
        expect(err instanceof NotOwner).toBeTruthy();
      }
    });
    it('should gracefully fail if the db fails', () => {
      try {
        expect.assertions(1);
        service.cancel(0, 5);
      } catch (err) {
        const invalid = new InvalidId();
        const expectedErr = new HttpException(
          invalid.message,
          HttpStatus.BAD_REQUEST,
        );
        expect(err).toMatchObject(expectedErr);
      }
    });
  });
});
