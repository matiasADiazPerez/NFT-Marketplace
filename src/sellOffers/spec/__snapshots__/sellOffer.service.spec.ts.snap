// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellOffersService find all sell offers should return all sell offers 1`] = `
[
  {
    "auctionState": 1,
    "createdAt": {},
    "currentBid": 0,
    "tokenId": 0,
    "updatedAt": {},
    "userId": 0,
  },
]
`;

exports[`SellOffersService find all sell offers should return all sell offers 3`] = `
{
  "auctionState": 1,
  "createdAt": {},
  "currentBid": 0,
  "tokenId": 0,
  "updatedAt": {},
  "userId": 0,
}
`;

exports[`SellOffersService find offers owned by user should return all sell offers of the user 1`] = `
[
  {
    "auctionState": 1,
    "createdAt": {},
    "currentBid": 0,
    "tokenId": 0,
    "updatedAt": {},
    "userId": 0,
  },
  {
    "DeletedAt": {},
    "auctionState": 1,
    "createdAt": {},
    "currentBid": 0,
    "tokenId": 0,
    "updatedAt": {},
    "userId": 0,
  },
]
`;

exports[`SellOffersService find one sell offers should return the corresponding sell offer 1`] = `
{
  "auctionState": 1,
  "createdAt": {},
  "currentBid": 0,
  "tokenId": 0,
  "updatedAt": {},
  "userId": 0,
}
`;
